src/system.rs

```diff
@@ -1357,6 +1426,93 @@ mod schema_system_tests {
         assert!(schema.is_ok());
     }
 
+    #[test]
+    fn schema_system_map_authority_with_invalid_inline_import_type_test() {
+        // map with (id, ion content)
+        let map_authority = [
+            (
+                "sample.isl",
+                r#"
+                    schema_header::{
+                      imports: [ { id: "sample_builtin_nullable_types.isl", type: my_text } ],
+                    }
+                    
+                    type::{
+                      name: my_type,
+                      type: my_text
+                    }
+                    
+                    schema_footer::{
+                    }
+                "#,
+            ),
+            (
+                "sample_builtin_nullable_types.isl",
+                r#"
+                    schema_header::{
+                      imports: [],
+                    }
+                    
+                    type::{
+                      name: my_text,
+                      one_of: [
+                        my_string, // this type reference was not improted in sample.isl
+                        symbol,
+                      ],
+                    }
+                    
+                    type::{
+                        name: my_string,
+                        type: string,
+                    }
+                    
+                    schema_footer::{
+                    }
+                "#,
+            ),
+        ];
+        let mut schema_system =
+            SchemaSystem::new(vec![Box::new(MapDocumentAuthority::new(map_authority))]);
+        // verify if the schema loads with an error for invalid type import,
+        // which has a dependency on a type reference that was not imported
+        let schema = schema_system.load_schema("sample.isl");
+        assert!(schema.is_err());
+    }
+
+    #[test]
+    fn schema_system_map_authority_with_multiple_type_definitions() {
+        // map with (id, ion content)
+        let map_authority = [(
+            "sample.isl",
+            r#"
+                    schema_header::{
+                      imports: [],
+                    }
+                    
+                    type::{
+                      name: my_text,
+                      one_of: [
+                        my_string, 
+                        symbol,
+                      ],
+                    }
+                    
+                    type::{
+                        name: my_string,
+                        type: string,
+                    }
+                    
+                    schema_footer::{
+                    }
+                "#,
+        )];
+        let mut schema_system =
+            SchemaSystem::new(vec![Box::new(MapDocumentAuthority::new(map_authority))]);
+        // verify if the schema loads without any errors
+        let schema = schema_system.load_schema("sample.isl");
+        assert!(schema.is_ok());
+    }
+
```

Comment:
Here are two more tests that I think cover important use cases for recursively defined data structures. I know that they are a little bit contrived (why would anyone have data that can only be alternately-nested lists and sexps?), but they are representative of recursive data types. The alternating recursive definition is a lot like ISL types and constraints, and the self-referencing recursion is basically a tree or heap.

```suggestion

    #[test]
    fn schema_system_map_authority_with_multiple_codependent_type_definitions() {
        // map with (id, ion content)
        let map_authority = [(
            "sample.isl",
            r#"
                    type::{
                       name: node_a,
                       type: list,
                       element: node_b,
                    }
                    
                    type::{
                       name: node_b,
                       type: sexp,
                       element: node_a,
                    }
                "#,
        )];
        let mut schema_system =
            SchemaSystem::new(vec![Box::new(MapDocumentAuthority::new(map_authority))]);
        // verify if the schema loads without any errors
        let schema = schema_system.load_schema("sample.isl");
        assert!(schema.is_ok());
    }

    #[test]
    fn schema_system_map_authority_with_self_referencing_type_definition() {
        // map with (id, ion content)
        let map_authority = [(
            "sample.isl",
            r#"
                    type::{
                       name: binary_heap_node,
                       type: sexp,
                       element: { one_of: [ binary_heap_node, int ] },
                       container_length: range::[0, 2],
                    }
                "#,
        )];
        let mut schema_system =
            SchemaSystem::new(vec![Box::new(MapDocumentAuthority::new(map_authority))]);
        // verify if the schema loads without any errors
        let schema = schema_system.load_schema("sample.isl");
        assert!(schema.is_ok());
    }

```

Reply:
I will add these tests with next commit